File: src/errorHandler.js
export class FlightAPIError extends Error {
  constructor(message, code = "GENERAL_ERROR", details = {}) {
    super(message);
    this.name = "FlightAPIError";
    this.code = code;
    this.details = details;
  }
}

export class ValidationError extends Error {
  constructor(message, field, value) {
    super(message);
    this.name = "ValidationError";
    this.field = field;
    this.value = value;
  }
}

export const errorCodes = {
  INVALID_AIRPORT_CODE: "INVALID_AIRPORT_CODE",
  INVALID_DATE_FORMAT: "INVALID_DATE_FORMAT",
  INVALID_PASSENGER_COUNT: "INVALID_PASSENGER_COUNT",
  FLIGHT_NOT_FOUND: "FLIGHT_NOT_FOUND",
  API_RATE_LIMIT: "API_RATE_LIMIT",
  API_UNAVAILABLE: "API_UNAVAILABLE",
  NETWORK_ERROR: "NETWORK_ERROR",
};

export function validateAirportCode(code) {
  if (!code || typeof code !== "string") {
    throw new ValidationError("Airport code is required", "airport_code", code);
  }
  
  if (code.length !== 3) {
    throw new ValidationError("Airport code must be 3 characters long", "airport_code", code);
  }
  
  if (!/^[A-Za-z]{3}$/.test(code)) {
    throw new ValidationError("Airport code must contain only letters", "airport_code", code);
  }
  
  return code.toUpperCase();
}

export function validateDate(dateString, fieldName = "date") {
  if (!dateString || typeof dateString !== "string") {
    throw new ValidationError(`${fieldName} is required`, fieldName, dateString);
  }
  
  const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
  if (!dateRegex.test(dateString)) {
    throw new ValidationError(
      `${fieldName} must be in YYYY-MM-DD format`,
      fieldName,
      dateString
    );
  }
  
  const date = new Date(dateString);
  if (isNaN(date.getTime())) {
    throw new ValidationError(`Invalid ${fieldName}`, fieldName, dateString);
  }
  
  // Check if date is not in the past (allow today)
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  date.setHours(0, 0, 0, 0);
  
  if (date < today) {
    throw new ValidationError(
      `${fieldName} cannot be in the past`,
      fieldName,
      dateString
    );
  }
  
  return dateString;
}

export function validateFlightNumber(flightNumber) {
  if (!flightNumber || typeof flightNumber !== "string") {
    throw new ValidationError("Flight number is required", "flight_number", flightNumber);
  }
  
  // Flight number format: 2-3 letter airline code + 1-4 digit number
  const flightRegex = /^[A-Za-z]{2,3}\d{1,4}$/;
  if (!flightRegex.test(flightNumber)) {
    throw new ValidationError(
      "Flight number must be in format: AA123, DL1234, etc.",
      "flight_number",
      flightNumber
    );
  }
  
  return flightNumber.toUpperCase();
}

export function validatePassengerCount(count) {
  if (count !== undefined) {
    if (!Number.isInteger(count) || count < 1 || count > 9) {
      throw new ValidationError(
        "Number of adults must be between 1 and 9",
        "adults",
        count
      );
    }
  }
  return count || 1;
}

export function validateTravelClass(travelClass) {
  const validClasses = ["economy", "premium_economy", "business", "first"];
  if (travelClass && !validClasses.includes(travelClass)) {
    throw new ValidationError(
      `Travel class must be one of: ${validClasses.join(", ")}`,
      "travel_class",
      travelClass
    );
  }
  return travelClass || "economy";
}

export function formatError(error) {
  console.error("Error occurred:", error);

  if (error instanceof ValidationError) {
    return {
      content: [
        {
          type: "text",
          text: `Validation Error: ${error.message}\\n\\nField: ${error.field}\\nValue: ${error.value}`,
        },
      ],
    };
  }

  if (error instanceof FlightAPIError) {
    let errorText = `Flight API Error: ${error.message}\\n\\nError Code: ${error.code}`;
    
    if (Object.keys(error.details).length > 0) {
      errorText += `\\n\\nDetails:\\n${JSON.stringify(error.details, null, 2)}`;
    }
    
    return {
      content: [
        {
          type: "text",
          text: errorText,
        },
      ],
    };
  }

  // Generic error handling
  return {
    content: [
      {
        type: "text",
        text: `An unexpected error occurred: ${error.message}\\n\\nPlease try again later or contact support if the issue persists.`,
      },
    ],
  };
}

File: flight_mcp_server/error_handler.py
"""Error handling and validation for flight MCP server."""

import re
from datetime import datetime, date
from typing import Any, Optional


class FlightAPIError(Exception):
    """Custom exception for flight API errors."""
    pass


class ValidationError(Exception):
    """Custom exception for validation errors."""
    pass


def validate_airport_code(code: Any) -> str:
    """Validate and normalize airport IATA code."""
    if not code or not isinstance(code, str):
        raise ValidationError("Airport code is required and must be a string")
    
    code = code.strip().upper()
    
    if not code.isalpha() or len(code) != 3:
        raise ValidationError("Airport code must be exactly 3 letters (e.g., LAX, JFK, LHR)")
    
    return code


def validate_date(date_str: Any, field_name: str = "date") -> str:
    """Validate date format (YYYY-MM-DD)."""
    if not date_str or not isinstance(date_str, str):
        raise ValidationError(f"{field_name} is required and must be a string")
    
    # Check format with regex
    if not re.match(r'^\d{4}-\d{2}-\d{2}$', date_str):
        raise ValidationError(f"{field_name} must be in YYYY-MM-DD format")
    
    # Validate actual date
    try:
        parsed_date = datetime.strptime(date_str, "%Y-%m-%d").date()
    except ValueError:
        raise ValidationError(f"Invalid {field_name}: {date_str}")
    
    # Check if date is not in the past (allow today)
    today = date.today()
    if parsed_date < today:
        raise ValidationError(f"{field_name} cannot be in the past")
    
    return date_str


def validate_passenger_count(count: Any) -> int:
    """Validate passenger count."""
    if count is None:
        return 1  # Default value
    
    if not isinstance(count, int):
        try:
            count = int(count)
        except (ValueError, TypeError):
            raise ValidationError("Passenger count must be a number")
    
    if count < 1 or count > 9:
        raise ValidationError("Passenger count must be between 1 and 9")
    
    return count


def validate_travel_class(travel_class: Any) -> str:
    """Validate travel class."""
    if travel_class is None:
        return "economy"  # Default value
    
    if not isinstance(travel_class, str):
        raise ValidationError("Travel class must be a string")
    
    travel_class = travel_class.lower().strip()
    valid_classes = ["economy", "premium_economy", "business", "first"]
    
    if travel_class not in valid_classes:
        raise ValidationError(f"Travel class must be one of: {', '.join(valid_classes)}")
    
    return travel_class


def validate_flight_number(flight_number: Any) -> str:
    """Validate flight number format."""
    if not flight_number or not isinstance(flight_number, str):
        raise ValidationError("Flight number is required and must be a string")
    
    flight_number = flight_number.strip().upper()
    
    # Flight number should be airline code (2-3 letters) + number
    if not re.match(r'^[A-Z]{2,3}\d{1,4}$', flight_number):
        raise ValidationError("Flight number must be in format like AA123, DL456, etc.")
    
    return flight_number


def validate_airline_code(airline_code: Any) -> str:
    """Validate airline IATA code."""
    if not airline_code or not isinstance(airline_code, str):
        raise ValidationError("Airline code is required and must be a string")
    
    airline_code = airline_code.strip().upper()
    
    if not airline_code.isalpha() or len(airline_code) not in [2, 3]:
        raise ValidationError("Airline code must be 2-3 letters (e.g., AA, DL, UA)")
    
    return airline_code